src/main/drivers/timer_def.h

```diff
@@ -699,4 +713,290 @@
 #define DEF_TIM_AF__PI6__TCH_TIM8_CH2     D(3, 8)
 #define DEF_TIM_AF__PI7__TCH_TIM8_CH3     D(3, 8)
 
+#elif defined(STM32H7)
+#define DEF_TIM(tim, chan, pin, flags, out, dmaopt, upopt) {            \
+    tim,                                                                \
+    IO_TAG(pin),                                                        \
+    DEF_TIM_CHANNEL(CH_ ## chan),                                       \
+    flags,                                                              \
+    (DEF_TIM_OUTPUT(CH_ ## chan) | out),                                \
+    DEF_TIM_AF(TCH_## tim ## _ ## chan, pin)                            \
+    DEF_TIM_DMA_COND(/* add comma */ ,                                  \
+        DEF_TIM_DMA_STREAM(dmaopt, TCH_## tim ## _ ## chan),            \
+        DEF_TIM_DMA_REQUEST(TCH_## tim ## _ ## chan),                   \
+        DEF_TIM_DMA_HANDLER(dmaopt, TCH_## tim ## _ ## chan)            \
+    )                                                                   \
+    DEF_TIM_DMA_COND(/* add comma */ ,                                  \
+        DEF_TIM_DMA_STREAM(upopt, TCH_## tim ## _UP),                   \
+        DEF_TIM_DMA_REQUEST(TCH_## tim ## _UP),                         \
+        DEF_TIM_DMA_HANDLER(upopt, TCH_## tim ## _UP)                   \
+    )                                                                   \
+}                                                                       \
+/**/
+
+#define DEF_TIM_CHANNEL(ch)                   CONCAT(DEF_TIM_CHANNEL__, DEF_TIM_CH_GET(ch))
+#define DEF_TIM_CHANNEL__D(chan_n, n_channel) TIM_CHANNEL_ ## chan_n
+
+#define DEF_TIM_AF(timch, pin)                CONCAT(DEF_TIM_AF__, DEF_TIM_AF_GET(timch, pin))
+#define DEF_TIM_AF__D(af_n, tim_n)            GPIO_AF ## af_n ## _TIM ## tim_n
+
+#define DEF_TIM_DMA_STREAM(variant, timch)                              \
+    CONCAT(DEF_TIM_DMA_STREAM__, DEF_TIM_DMA_GET(variant, timch))
+#define DEF_TIM_DMA_STREAM__D(dma_n, stream_n)  DMA ## dma_n ## _Stream ## stream_n
+#define DEF_TIM_DMA_STREAM__NONE                        NULL
+
+// XXX This is awful. There must be some smart way of doing this ...
+#define DEF_TIM_DMA_REQUEST(timch) \
+    CONCAT(DEF_TIM_DMA_REQ__, DEF_TIM_TCH2BTCH(timch))
+
+#define DEF_TIM_DMA_HANDLER(variant, timch) \
+    CONCAT(DEF_TIM_DMA_HANDLER__, DEF_TIM_DMA_GET(variant, timch))
+#define DEF_TIM_DMA_HANDLER__D(dma_n, stream_n) DMA ## dma_n ## _ST ## stream_n ## _HANDLER
+#define DEF_TIM_DMA_HANDLER__NONE                       0
+
+/* H7 Stream Mappings */
+// D(DMAx, Stream)
+
+// H7 has DMAMUX that allow arbitrary assignment of peripherals to streams.
```

Comment:
It may be better to refactor this into code in future ...

Reply:
Yeap.